plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java

```diff
@@ -838,6 +863,119 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
                         "com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: first attempt is poised to fail");
     }
 
+    @Test
+    public void testSpecificSchemaAndTableCachesWithCacheMissing()
+            throws Exception
+    {
+        assertSpecificSchemaAndTableCaches(true);
+    }
+
+    @Test
+    public void testSpecificSchemaAndTableCachesWithoutCacheMissing()
+            throws Exception
+    {
+        assertSpecificSchemaAndTableCaches(false);
+    }
+
+    private void assertSpecificSchemaAndTableCaches(boolean cacheMissing)
+    {
+        CachingJdbcClient cachingJdbcClient = createCachingJdbcClient(
+                FOREVER,
+                Optional.of(Duration.succinctDuration(2, SECONDS)),
+                Optional.of(Duration.succinctDuration(1, SECONDS)),
+                cacheMissing,
+                10000);
+        ConnectorSession session = createSession("asession");
+        SchemaTableName firstName = new SchemaTableName(schema, "first_table");
+        SchemaTableName secondName = new SchemaTableName(schema, "second_table");
+
+        JdbcTableHandle first = createTable(firstName);
+        assertSchemaNamesCache(cachingJdbcClient).loads(1).misses(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getSchemaNames(session)).contains(schema);
+        });
+        assertTableNamesCache(cachingJdbcClient).loads(1).misses(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableNames(session, Optional.empty())).contains(firstName);
+        });
+        assertTableHandleCache(cachingJdbcClient).misses(1).loads(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+        });
+        assertTableHandleCache(cachingJdbcClient).misses(1).loads(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableHandle(session, secondName)).isEmpty();
+        });
+
+        JdbcTableHandle second = createTable(secondName);
+        assertSchemaNamesCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getSchemaNames(session)).contains(schema);
+        });
+        assertTableNamesCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableNames(session, Optional.empty()))
+                    .contains(firstName)
+                    .doesNotContain(secondName);
+        });
+        assertTableHandleCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+        });
+        if (cacheMissing) {
+            assertTableHandleCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, secondName)).isEmpty();
+            });
+        }
```

Comment:
Can we have dedicated test with cache disabled ?

Reply:
The difference is pretty small, but the method is large. It will be hard to spot differences if there will be two different flavors of it.